Add SecureBoot support for arbitrary operating systems to "Grub2 UEFI" PXE loaders (#2145)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-server-to-provision-secure-boot-enabled-hosts.adoc

[options="nowrap" subs="+quotes"]
----
# install -o foreman-proxy -g foreman-proxy -d /usr/local/share/bootloader-universe/_My_OS_Dependent_Directory_Name_/
----
. Download the shim and GRUB2 packages for the operating system of your target host:
+
[options="nowrap" subs="+quotes"]
----
# wget -O /tmp/{grub_efi_package_name} {grub_efi_download_url}{grub_efi_package_name}
# wget -O /tmp/{shim_efi_package_name} {shim_efi_download_url}{shim_efi_package_name}
----
. Extract the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes"]
----
# {extract_grub}
# {extract_shim}
----
. Make the shim and GRUB2 binaries available for host provisioning:
+
[options="nowrap" subs="+quotes"]
----
# cp {grub_binary_path} /usr/local/share/bootloader-universe/_My_OS_Dependent_Directory_Name_/grubx64.efi
# cp {shim_binary_path} /usr/local/share/bootloader-universe/_My_OS_Dependent_Directory_Name_/shimx64.efi
----
. Set the permissions for the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes"]
----
# chmod 444 /usr/local/share/bootloader-universe/_My_OS_Dependent_Directory_Name_/grubx64.efi
# chmod 444 /usr/local/share/bootloader-universe/_My_OS_Dependent_Directory_Name_/shimx64.efi
----

> You can now provision Secure Boot enabled {client_os} hosts with the *Grub2 UEFI SecureBoot (target OS)* PXE loader.

Reviewer: ```suggestion
You can now provision Secure Boot enabled {client-os} hosts with the *Grub2 UEFI SecureBoot (target OS)* PXE loader.
```

----
# wget -O /tmp/{grub_efi_downloaded_package_name} _https://{server-example-com}/{grub_efi_downloaded_package_name}_
# wget -O /tmp/{shim_efi_downloaded_package_name} _https://{server-example-com}/{shim_efi_downloaded_package_name}_
----
+
You can download the `{grub_efi_package_name}` package from {grub_efi_download_url}.
You can download the `{shim_efi_package_name}` package from {shim_efi_download_url}.
. Extract the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# {extract_grub}
# {extract_shim}
----
. Make the shim and GRUB2 binaries available for host provisioning:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# cp {grub_efi_tmp_binary_path} _My_Bootloader_Directory_/{client-os-context}/grubx64.efi
# cp {shim_efi_tmp_binary_path} _My_Bootloader_Directory_/{client-os-context}/shimx64.efi
----
. Set the permissions for the shim and GRUB2 binaries:
+
[options="nowrap" subs="+quotes,verbatim,attributes"]
----
# chmod 644 _My_Bootloader_Directory_/{client-os-context}/grubx64.efi
# chmod 644 _My_Bootloader_Directory_/{client-os-context}/shimx64.efi
----
ifeval::["{client-pkg-ext}" == "deb"]

> . Link the `grub.cfg` file from the TFTP servers `grub2` folder to the legacy `grub` folder:

Reviewer 2: ```suggestion
. Link the `grub.cfg` file from the TFTP servers `grub2` folder to the legacy `grub` folder:
+
```

[id="Configuring-{project-context}-Server-to-Provision-{client-os-context}-on-Secure-Boot-Enabled-Hosts_{context}"]

> = Configuring {ProjectServerTitle} to Provision {client-os} on Secure Boot Enabled Hosts

Reviewer 2: ```suggestion
= Configuring {ProjectServerTitle} to provision {client-os} on Secure Boot enabled hosts
```

Secure Boot follows a chain of trust from the start of the host to the loading of Linux kernel modules.

> The first shim that is loaded determines which distribution is allowed to be booted or kexec'ed until next reboot.

Reviewer 2: Unsure if "kexec'ed" is allowed by IBM style guide. I am OK, any concerns? @asteflova ?

Reviewer 3: IBM style guide is not that detailed but it generally does not allow turning commands into verbs.

Maybe this would work: "The first shim that is loaded determines which distribution can be booted or loaded by using a `kexec` system call until the next reboot."

Reviewer: IMO your suggestion would fit, @apinnick.

What do you think, @jloeser?

Author: Would be OK.

Reviewer: Reworded

[IMPORTANT]
====

> For each subnet you have to perform the following steps on the corresponding Smart Proxy that is configured as a TFTP proxy to be able to provision {client-os} on Secure Boot enabled hosts using the *Grub2 UEFI SecureBoot (target OS)* PXE loader.

Reviewer 2: ```suggestion
You have to perform the following steps on your {SmartProxy} that is configured as a TFTP proxy to provision {client-os} on Secure Boot enabled hosts.
```

alternative:

```suggestion
You have to perform the following configuration steps on each {SmartProxy} used as a TFTP proxy for the subnet of your Secure Boot enabled host.
```

Reviewer: Thanks for your suggestions, @maximiliankolb. I kind of combined them both. What do you think:
```suggestion
You have to perform the following configuration steps on each {SmartProxy} used as a TFTP proxy for a subnet to provision Secure Boot enabled hosts on it.
```

====

> .Prerequisite

Reviewer 3: ```suggestion
.Prerequisites
```
IBM Style Guide says to use the plural even if there is only one prerequisite.

ifeval::["{client-os}" == "Debian"]
* Ensure that `ar` and `xz` are installed on your {ProjectServer}.
endif::[]
ifeval::["{client-os}" == "Ubuntu"]
* Ensure that `ar`, `xz`, and `zstd` are installed on your {ProjectServer}.
endif::[]
ifeval::["{client-pkg-ext}" == "rpm"]
* Ensure that `cpio` is installed on your {ProjectServer}.
endif::[]

.Procedure

> . On your {ProjectServer}, configure the directory where you want to store the GRUB2 and shim binaries required for provisioning Secure Boot enabled hosts:

Reviewer 2: ```suggestion
. On your {ProjectServer}, configure the directory to store the GRUB2 and shim binaries required for provisioning Secure Boot enabled hosts:
```
